modules/imgproc/src/demosaicing.cpp

```diff
@@ -538,6 +538,233 @@ class SIMDBayerInterpolator_8u
         return 0;
     }
 };
+#elif CV_MSA
+class SIMDBayerInterpolator_8u
+{
+public:
+    SIMDBayerInterpolator_8u()
+    {
+    }
+
+    int bayer2Gray(const uchar* bayer, int bayer_step, uchar* dst,
+                   int width, int bcoeff, int gcoeff, int rcoeff) const
+    {
+        /*
+         B G B G | B G B G | B G B G | B G B G
+         G R G R | G R G R | G R G R | G R G R
+         B G B G | B G B G | B G B G | B G B G
+         */
+        v8u16 masklo = msa_dupq_n_u16(255);
+        const uchar* bayer_end = bayer + width;
+
+        for( ; bayer <= bayer_end - 18; bayer += 14, dst += 14 )
+        {
+            v8u16 r0 = msa_ld1q_u16((const ushort*)bayer);
+            v8u16 r1 = msa_ld1q_u16((const ushort*)(bayer + bayer_step));
+            v8u16 r2 = msa_ld1q_u16((const ushort*)(bayer + bayer_step * 2));
```

Comment:
Main principle of "universal intrinsics" (SIMD) approach is "code once, run anywhere".

You need to write code once (on some platform), debug it with some backend (there is even C++ backend for SIMD for reference) and then it should run on other backends with reasonable speed.

So we are strictly against huge code blocks using native intrinsic if this code can be replaced by universal SIMD code without significant lost of speed.
Universal SIMD code is regularly tested via x86/ARM backends, native RAW MIPS has very low testing coverage.

See discussion here: https://github.com/opencv/opencv/pull/9763#issuecomment-335159324

Could you please move these changes into a separate PR?
Lets merge first addition of SIMD backend and related build scripts (so we can add cross-platforms builds for that into our infrastructure (but no tests)).